### PR TITLE
Laboratory work #3, Elena Ermak - 22FPL1

### PR DESCRIPTION
**Pull request by ermakelena**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def __init__(self, end_of_word_token: str) -> None:
         Args:
             end_of_word_token (str): A token denoting word boundary
         """
+        self._end_of_word_token = end_of_word_token
+        self._storage = {"_": 0}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:27]: Strange constant

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or not text or not any(char.isalpha() for char in text):
+            return None
+        self.text = text
+        splited_text = text.split()
+        final_text = []
+        for word in splited_text:
+            clear_word = "".join(char.lower() for char in word if char.isalpha())
+            clear_word = list(clear_word)
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:53]: join and then list... too much - simplify

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -76,6 +99,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        inv_storage = {value: key for key, value in self._storage.items()}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:105]: Try to do this without creating a new storage

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -104,6 +150,19 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element in self._storage:
+            return self._storage[element]
+        key = len(self._storage)
+        while key <= len(self._storage):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:158]: Very strange implementation

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -159,6 +243,18 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+        decoded_corpus = list(decoded_corpus)
+        if decoded_corpus[-1] == "_":
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:249]: again strange constant

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:27]: fixed

Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:27]: no

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py]: I fixed it locally before pushing, now the code is different

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:54]: simplified

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:106]: done

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:158]: changed it

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py]: fixed it

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -3,6 +3,8 @@
 
 Beam-search and natural language generation evaluation
 """
+#import json
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:6]: ?

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -41,6 +45,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         """
+        if not isinstance(text, str) or not text or not any(char.isalpha() for char in text):
+            return None
+
+        splited_text = text.split()
+        final_text = []
+        for word in splited_text:
+            clear_word = [char.lower() for char in word if char.isalpha()]
+            if clear_word:
+                final_text += clear_word
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:56]: extend

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -76,6 +100,14 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+
+        for key, id_num in self._storage.items():
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:106]: filter

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -104,6 +150,12 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:155]: early return

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -208,6 +300,21 @@ def build(self) -> int:
...
+            return 1
+
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+
+        if n_grams is None:
+            return 1
+
+        for n_gram in set(n_grams):
+            absolute_frequency = n_grams.count(n_gram)
+            beginning_frequency = len([n_gram_begin for n_gram_begin in n_grams
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:313]: count

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -222,8 +329,23 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
...
+            if context == n_gram[:self._n_gram_size - 1]:
+                letter = n_gram[self._n_gram_size - 1]
+                if letter not in next_letter:
+                    next_letter[letter] = freq
+
+        return next_letter
+
     def _extract_n_grams(
-        self, encoded_corpus: tuple[int, ...]
+            self, encoded_corpus: tuple[int, ...]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:348]: revert

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -310,11 +475,25 @@ def get_next_token(self, sequence: tuple[int, ...]) -> Optional[list[tuple[int,
...
+
+        return sorted(list(next_tokens.items()),
+                             key=lambda x: (-x[1], -x[0]))[:self._beam_width]
+
     def continue_sequence(
-        self,
-        sequence: tuple[int, ...],
-        next_tokens: list[tuple[int, float]],
-        sequence_candidates: dict[tuple[int, ...], float],
+            self,
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:493]: revert

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -332,8 +511,21 @@ def continue_sequence(
...
+        for token in next_tokens:
+            seq = sequence + (token[0],)
+            sequence_candidates[seq] = sequence_candidates[sequence] - math.log(token[1])
+        del sequence_candidates[sequence]
+
+        return sequence_candidates
+
     def prune_sequence_candidates(
-        self, sequence_candidates: dict[tuple[int, ...], float]
+            self, sequence_candidates: dict[tuple[int, ...], float]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:528]: reert

File: lab_3_generate_by_ngrams/main.py [hunk 14]
```diff
@@ -460,9 +704,9 @@ class BackOffGenerator:
     """
 
     def __init__(
-        self,
-        language_models: tuple[NGramLanguageModel, ...],
-        text_processor: TextProcessor,
+            self,
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:707]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:56]: done

File: lab_3_generate_by_ngrams/main.py (hunk 8 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:106]: done

File: lab_3_generate_by_ngrams/main.py (hunk 9 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:155]: done i guess

File: lab_3_generate_by_ngrams/main.py (hunk 10 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py]: done

File: lab_3_generate_by_ngrams/main.py [hunk 15]
```diff
@@ -208,6 +308,27 @@ def build(self) -> int:
...
+        if n_grams is None:
+            return 1
+
+        beginning_freq_dict = {}
+
+        for n_gram in set(n_grams):
+            absolute_frequency = n_grams.count(n_gram)
+            beginning = n_gram[:-1]
+            if beginning not in beginning_freq_dict:
+                beginning_freq_dict[beginning] = [n_gram_begin[:-1] == beginning
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:325]: too hard to understand, please rework

File: lab_3_generate_by_ngrams/main.py (hunk 11 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:362]: agree, Artyom

File: lab_3_generate_by_ngrams/main.py (hunk 12 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:507]: again agree

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 15 above)
Comment (ermakelena) [reply lab_3_generate_by_ngrams/main.py:325]: reworked

**Pull request closed without merging**